File: fnc_play_and_record.py
import sounddevice as sd
import numpy as np


def fnc_play_and_record(device=None, Fs=48000, audio_playback=None, channel_playback=None, channel_receive=None):
    """
    Function to playback and record audio, with the ability to insert a voice command.

    Args:
        device (str, optional): Name of the audio device to use. Defaults to None.
        Fs (int, optional): Sampling frequency. Defaults to 48000.
        audio_playback (numpy.ndarray, optional): Input audio data for playback. Defaults to None.
        channel_receive (list, optional): Input channels to record from. Defaults to None.
        channel_playback (list, optional): Output channels to play back to. Defaults to None.
    Returns:
        tuple: Recorded audio data and the final index.
    """

    if device is None and audio_playback is None:
        flag_info_only = True
    else:
        flag_info_only = False

    if device is not None and audio_playback is None:
        flag_dev_info = True
    else:
        flag_dev_info = False

    if audio_playback is not None:
        assert audio_playback.shape[1] == len(channel_playback), ("Row dimension of the playback signal should match "
                                                                  "the number of playback channels")

    # Display information about audio devices on system
    if flag_info_only:
        print("The following devices were recognized:")
        info_dev = sd.query_devices()
        for i, dev in enumerate(info_dev):
            print(f"     {i}: {dev['name']}")
        print("Your device is probably the one with the longest name :)")
        print(" ")
        return

    # Either give device-specific info or playback and record audio
    if not flag_info_only:
        info_dev = sd.query_devices()
        print(f"The specified device is: {device}. Testing to see if it exists...")

        # Select user-specified device as playback device
        success_flag = False
        for dev_id, dev in enumerate(info_dev):
            if device in dev['name']:
                success_flag = True
                break

        if success_flag:
            print(f"The selected device is: {info_dev[dev_id]['name']}")
            if flag_dev_info:
                print("Some information on the selected device:")
                print(sd.query_devices(dev_id))
                print(" ")
        else:
            print("The specified device was not found for the given sample frequency. ")
            print("The following devices were recognized:")
            info_dev = sd.query_devices()
            for i, dev in enumerate(info_dev):
                print(f"     {i}: {dev['name']}")
            print(" ")
            return None

        # Adjust output channels to the device's max output channels if needed
        if not flag_dev_info and success_flag:
            terminate_flag = False
            max_output_channels = info_dev[dev_id]['max_output_channels']
            if max(channel_playback) > max_output_channels:
                print(f"The device supports up to {max_output_channels} playback channels. A larger one was specified. Terminating.")
                terminate_flag = True

            if max(channel_receive) > info_dev[dev_id]['max_input_channels']:
                print(
                    f"The device supports up to of {info_dev[dev_id]['max_input_channels']} receiver channels. A larger one was specified. Terminating.")
                terminate_flag = True

            if not terminate_flag:
                # Determine the maximum value between the two channel arrays
                max_in = np.max(channel_receive)
                max_out = np.max(channel_playback)
                max_channel = max(max_in, max_out)   # Note that max_channel is also the number of channels

                # Create an array of channel indices based on the max
                channel_mix = np.arange(0, max_channel)

                # Placeholder for recorded audio
                audio_out = np.zeros((audio_playback.shape[0], len(channel_receive)), dtype=np.float32)

                # Playback and record audio
                print(f"Note: your audio duration is {len(audio_playback) / Fs} seconds")
                buf_size = 512
                buf_default = np.zeros((buf_size, len(channel_mix)), dtype='float32')  # The default playback audio is all zeros.
                i = 0
                try:
                    # Note that sd.Stream does not take channels, but instead takes a number of channels (!!!!)
                    # Hence, in the code above, we need to make the number of channels and the playback audio match the
                    # target channels as specified by the user
                    with sd.Stream(device=dev_id, samplerate=Fs, channels=max_channel, dtype='float32') as stream:
                        # Loop indefinitely until the program is terminated
                        while i + buf_size < len(audio_playback):
                            end_index = i + buf_size

                            # Set the buffer: all-zeros apart from the channels containing the playback signals.
                            buf = buf_default
                            buf[:, channel_playback-1] = audio_playback[i:end_index, :]  # The offset by 1 is due to counting from zero

                            stream.write(buf)
                            indata = stream.read(end_index - i)[0]
                            audio_out[i:end_index] = indata[:, channel_receive-1]  # The offset by 1 is due to counting from zero

                            i += buf_size

                except Exception as e:
                    print(f"An error occurred: {e}")
                    return None

                return audio_out
            else:
                return None

File: MATLAB/fnc_play_and_record.m
%Date:      November 2023
%Last edit: September 2024
%Author:    Dimme
%Descr:     This function can be used to simultaneously read and write, audio to/from a given soundcard. For example an RME Fireface.
%           the function does not do checks on the input etc, but does say if your soundcard is not found.
%
%Based on:  https://nl.mathworks.com/help/audio/ref/audioplayerrecorder-system-object.html
%
%Inputs:    device          a string describing the device. I.e. if your device is "USB CARD: RME FIREFACE", device="USB" (or any other part uniquely describing the device name). 
%           Fs              the desired sampling frequency in Hz. Note that the valid inputs depend on the device
%           audioPlayback   a LENGHT x N_Playback_channel matrix of audio. This is the audio that will be played back. It might be required to ensure the largest absolute value is 1, though I'm not entirely sure.
%           channelPlayback The channels on which the loudspeakers are connected.
%           channelReceive  The channels on which the microphones are connected.
%
%Outputs:   audioOut        A LENGTH x N_receiver_channel containing the microphone signals
%
%Notes:     - If you dont want to playback anything, just set it all zero and select an arbitrary but valid set of playback channels
%           - If you dont want to record anything, dont assing the output argument. For the receiver channels, select an arbitrary but valid set. 
%           - In my experience, not all channel assignments are allowed. For example, I once required channel 15:22 but this threw an error. Using 15:24 worked for some reason. 
%           - you can use fnc_play_and_record(). This will list the available devices.
%           - You can use fnc_play_and_record(device) or fnc_play_and_record(device, Fs). This will give you device info. 
%           - If your audioPlayback matrix does not match the channelPlayback parameter, theplayback matrix will either be truncated or zeros will be appended!
%           - In my experience, you sometimes need to restart MATLAB for the device list to update. 

function varargout = fnc_play_and_record(device, Fs, audioPlayback, channelPlayback, channelReceive)
    if nargin == 0
        flag_info_only = true;
    else
        flag_info_only = false;
    end
    if nargin == 1 
        flag_dev_info = true;
        Fs = 48000;
    elseif nargin==2
        flag_dev_info = true;
    else
        flag_dev_info = false;
    end
    if nargin == 3
        channelReceive =  [1,2];  %Recorder channels
        channelPlayback = [1,2];  %Playback channels
    end
    if nargin == 4
        channelReceive =  [1,2];  %Recorder channels
    end

    %Give user some information about audiodevices on system
    if flag_info_only 
        %Create audioPlayerRecorder object for the specified sampling frequency. I think this also accepts invalid sampling frequencies 
        playRec = audioPlayerRecorder;
        infoDev = getAudioDevices(playRec);
        disp("The following devices were recognised:")
        for i=1:length(infoDev)
            disp("     " + infoDev{i})
        end
        disp("Your device is probably the one with the longest name :)")
        disp(" ")
    end

    %Either give device specific onfo or playback and record audio
    if ~flag_info_only
        playRec = audioPlayerRecorder(Fs);
        infoDev = getAudioDevices(playRec);
        disp("The specified device is: " + device + ". testing to see if it exists...")
    
        %Select  user specified device as playback device
        for devId=1:length(infoDev)
            if isempty(strfind(infoDev{devId},device))~=1
                succes_flag = 1;
                break
            end
            succes_flag = 0;
        end
        if succes_flag==1
            disp("The selected device is: " + infoDev{devId}) 
            release(playRec)
            playRec.Device = infoDev{devId};  %update the playRec object to be the selected device
            if flag_dev_info                    %Give device specific info
                infoDevSum = info(playRec);
                disp("Some information on the selected device:")
                disp(infoDevSum);
            end
        else
            disp("The specified device was not found for the given sample frequency. Try running <fnc_play_and_record> without in- and output argumens")
        end

        %Playback audio
        if ~flag_dev_info && succes_flag == 1
            disp("Note: your audio duration is " + num2str(size(audioPlayback,1)/Fs) + " seconds");
            playRec.PlayerChannelMapping = channelPlayback;      
            playRec.RecorderChannelMapping = channelReceive;   

            %check if the specified audio has the right nuimber of channels. If not, add zeros or truncate and give warning
            if size(audioPlayback,2)>length(channelPlayback)
                disp("This device does not have sufficiently large number of playback channels for your audio. Truncating")
                audioPlayback = audioPlayback(:, 1:length(channelPlayback));
            elseif size(audioPlayback,2) < length(channelPlayback)
                disp("This device does have more playback channels than your audio input. Adding zeros")
                Nchan_audio = size(audioPlayback,2);
                Nchan = length(channelPlayback);
                audioPlayback = [audioPlayback, zeros(size(audioPlayback,1), Nchan-Nchan_audio)];
            end
            
            %Placeholder for recorded audio. 
            audioOut = zeros(size(audioPlayback,1), length(channelReceive));                

            %(4) Playback and record audio 
            buf = dsp.AsyncBuffer(size(audioPlayback,1));
            write(buf,audioPlayback);
            Indx = 0;
   
            % I read/write 512 samples. There probably are cases where it makes sense to change this (real-time processing?), though i'm not sure. 
            % Anyway, note that you can also add processing steps in this loop if they are quick enough. 
            nSample = 512;
            while buf.NumUnreadSamples >= nSample
                audioToPlay = read(buf,nSample);
                [audioRecorded,nUnderruns,nOverruns] = playRec(audioToPlay);

                audioOut(Indx+1:Indx+nSample, :) = audioRecorded;
                if nUnderruns > 0
                    fprintf('Audio player queue was underrun by %d samples.\n',nUnderruns);
                end
                if nOverruns > 0
                    fprintf('Audio recorder queue was overrun by %d samples.\n',nOverruns);
                end
                Indx = Indx+nSample;
            end
            release(buf)
            release(playRec)
        
        else
            audioOut = [];
        end
        varargout{1} = audioOut;
    end
end
